### ISSUE TITLE
[bug] 'assert node.pref.timestamp is not None' w/ test_requires in Conan v2 beta

### ISSUE THREAD

**Issue by ssrobins** [conan-io/conan#11606]
### Environment Details (include every applicable attribute)
  * Operating System+version: macOS 12.3 (Monterey)
  * Compiler+version: AppleClang 13.1.6.13160021
  * Conan version: 2.0.0-beta1
  * Python version: Python 3.10.5

### Steps to reproduce (Include if Applicable)
1. `git clone https://github.com/ssrobins/stackblox.git`
1. `cd stackblox`
1. `git checkout conan_v2_crash_gtest`
1. Delete ~/.conan2
1. `cmake --preset macos`

### Logs (Executed commands with output) (Include/Attach if Applicable)
The conan install command embedded in CMake fails with this error:
```
bzip2/1.0.8@ssrobins: Retrieving package 24df4bc901e0688f1084351da5b2fa6f658304da from remote 'artifactory-ssrobins' 
Downloading conanmanifest.txt
Downloading conaninfo.txt
Downloading conan_package.tgz
Decompressing conan_package.tgz
bzip2/1.0.8@ssrobins: Package installed 24df4bc901e0688f1084351da5b2fa6f658304da
bzip2/1.0.8@ssrobins: Downloaded package revision 14f612c81887fa9f529df75848f822cf
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/cli/cli.py", line 163, in run
    command.run(self._conan_api, self._commands[command_argument].parser, args[0][1:])
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/cli/command.py", line 157, in run
    info = self._method(conan_api, parser, *args)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/cli/commands/install.py", line 170, in install
    conan_api.install.install_binaries(deps_graph=deps_graph, remotes=remote, update=args.update)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conan/api/subapi/__init__.py", line 21, in wrapper
    return f(subapi, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conan/api/subapi/install.py", line 29, in install_binaries
    installer.install(deps_graph)
ERROR: 
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/client/installer.py", line 221, in install
    self._download_bulk(install_order)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/client/installer.py", line 249, in _download_bulk
    self._download_pkg(node)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/conans/client/installer.py", line 254, in _download_pkg
    assert node.pref.timestamp is not None
AssertionError
```

The key to the crash is switching the gtest package from `requires` to `test_requires`:
https://github.com/ssrobins/conan-recipes/commit/c2bf54494ddbc618032ced9ad5615fcf703c996e
...which I obviously want so my app doesn't get built with gtest.

I wasn't seeing this in local builds, but it was happening in GitHub Actions CI. Deleting ~/.conan2 let me reproduce it locally. Happens in Windows, but not Linux CI environments.

Reverting this commit and repeating the steps above prevents the error:
https://github.com/ssrobins/stackblox/commit/bf26cd602a5815bfb67e5c5e4be8d4b7f695ee36

**Comment by memsharded** [comment]
Hi @ssrobins 

I have been having a look, and while I have one possible point that could be causing this, I haven't been able to reproduce in a test. I haven't been able to reproduce with your example, as it has too many moving pieces.

Can you please try running your project applying this patch to the source code?

```python
diff --git a/conans/client/graph/graph_binaries.py b/conans/client/graph/graph_binaries.py
index a54933950..72f398ff6 100644
--- a/conans/client/graph/graph_binaries.py
+++ b/conans/client/graph/graph_binaries.py
@@ -89,6 +89,7 @@ class GraphBinariesAnalyzer(object):
             node.binary = previous_node.binary
             node.binary_remote = previous_node.binary_remote
             node.prev = previous_node.prev
+            node.pref_timestamp = previous_node.pref_timestamp

             # this line fixed the compatible_packages with private case.
             # https://github.com/conan-io/conan/issues/9880
```

Please let me know, many thanks!

**Comment by ssrobins** [comment]
Yup, that change fixed it! Sorry I didn't have a simpler setup to repro the problem. If you need something to further confirm this fix, let me know, I'll have some time Wednesday (USA evening) to work on it.

**Comment by memsharded** [comment]
Don't worry, and thanks very  much in any case for the detailed repro case. Even if I was unable to fully reproduce, it contained many good hints, and I found the bug fix thanks to it 😄 

No need to do more, thanks, I will now keep trying to reproduce in a test, and in the worst case, I will submit the PR with the fix only.

**Comment by memsharded** [comment]
Quick question: is it failing absolutely always? Is it deterministic? You are saying ``Happens in Windows, but not Linux CI environments.``, but the report details says it is OSX. Could you please clarify this? Thanks!

**Comment by ssrobins** [comment]
Yeah, as long as I delete ~/.conan2, it always happens.

GitHub Actions CI is using [ubuntu-20.04](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md) to build Android, [macos-12](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-12-Readme.md) to build macOS and iOS, [ubuntu-18.04](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-Readme.md) to build Linux, and [windows-2022](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md) to build Windows.

I misspoke about it not failing on linux, those were results from a different branch. When I re-ran the CI on the `conan_v2_crash_gtest` branch three times, ALL platforms failed with the same `assert node.pref.timestamp is not None` error.

Here's one of the runs, but I'm not sure if GitHub shows the log output to everyone:
https://github.com/ssrobins/stackblox/actions/runs/2657620709
If not, you could fork the repo and run it yourself. Everything is self-contained in GitHub Actions.

**Comment by ssrobins** [comment]
Also, the reason I concentrated on macOS in the report was because I was doing the dev work on a macbook so it was most convenient to try things out on that platform, while relying on GitHub Actions for the rest of them.

**Comment by memsharded** [comment]
Thanks for the details. Still unable to reproduce with a unit-test, but don't worry, at least the PR is open with the fix.

**Comment by memsharded** [comment]
Good news, I didn't give up, and I managed to make a repro unit-test! It was a very subtle detail of multiple ``test_requires``, and some being skipped as not used, and some being used.

Now the PR is really ready for review and beta.2

**Comment by ssrobins** [comment]
Nice work! Looking forward to beta.2!

**Comment by memsharded** [comment]
Closed by https://github.com/conan-io/conan/pull/11619, to be released in beta.2

**Comment by ssrobins** [comment]
Confirmed that the bug is still fixed in 2.0.0-beta2. Deleted my conan_v2_crash_gtest since this is fixed and the work has been merged to main.
